x-pack/plugins/code/public/components/hover/hover_widget.tsx

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { EuiButton, EuiFlexGroup, EuiText } from '@elastic/eui';
+// @ts-ignore
+import { renderMarkdown } from 'monaco-editor/esm/vs/base/browser/htmlContentRenderer';
+// @ts-ignore
+import { tokenizeToString } from 'monaco-editor/esm/vs/editor/common/modes/textToHtmlTokenizer';
+import React from 'react';
+import { MarkedString } from 'vscode-languageserver-types';
+
+export interface HoverWidgetProps {
+  state: HoverState;
+  contents?: MarkedString[];
+  fontFamily?: string;
+  gotoDefinition: () => void;
+  findReferences: () => void;
+}
+
+export enum HoverState {
+  LOADING,
+  INITIALTING,
+  READY,
+}
+
+export class HoverWidget extends React.PureComponent<HoverWidgetProps> {
+  public render() {
+    let contents;
+
+    switch (this.props.state) {
+      case HoverState.READY:
+        contents = this.renderContents();
+        break;
+      case HoverState.INITIALTING:
+        contents = this.renderInitialting();
+        break;
+      case HoverState.LOADING:
+      default:
+        contents = this.renderLoading();
+    }
+    return (
+      <div className="monaco-editor-hover-content">
+        {contents}
+        <EuiFlexGroup className="button-group euiFlexGroup" gutterSize="none" responsive={true}>
+          <EuiButton size="s" onClick={this.props.gotoDefinition}>
+            Goto Definition
+          </EuiButton>
+          <EuiButton size="s" onClick={this.props.findReferences}>
+            Find Reference
+          </EuiButton>
+          <EuiButton size="s">Go to Type</EuiButton>
+        </EuiFlexGroup>
+      </div>
+    );
+  }
+
+  private renderLoading() {
+    return (
+      <div className="hover-row">
+        <div className="text-placeholder gradient" />
+        <div className="text-placeholder gradient" style={{ width: '90%' }} />
+        <div className="text-placeholder gradient" style={{ width: '75%' }} />
+      </div>
+    );
+  }
+
+  private renderContents() {
+    return this.props
+      .contents!.filter(content => !!content)
+      .map((markedString, idx) => {
+        let markdown: string;
+        if (typeof markedString === 'string') {
+          markdown = markedString;
+        } else if (markedString.language) {
+          markdown = '```' + markedString.language + '\n' + markedString.value + '\n```';
+        } else {
+          markdown = markedString.value;
+        }
+        const renderedContents: string = renderMarkdown(
```

Comment:
When using `markdown-it` we also ensure that user's are unable to provide "raw HTML" and are locked down to the explicit markdown syntax. Additionally, we'll want to ensure that  monaco's markdown editor is safe to use with "user defined templates" and doesn't result in arbitrary code execution, as this can lead to an XSS.

Reply:
@legrego @kobelb we thought vscode's markdown rendering should be fairly battle tested, if that has security issue then, MS' own website and all other websites uses Monaco should be attackable by this method

in addition, the `renderMarkdown` code is here:
https://github.com/Microsoft/vscode/blob/dc261e04364542f6114d0e838b7fedf18ee505b7/src/vs/base/browser/htmlContentRenderer.ts

it's fairly simple only render text, links, image, language server only returns text and links, and our code disable rendering link as well, so it should be safe to use.

Also, we prefer keep vscode markdown in this particular place so that the style could be consistent.

Reply:
Thanks for the explanation - I'm ok with this approach